internal/clients/aws.go

```diff
@@ -25,39 +23,26 @@ import (
 )
 
 const (
-	// Terraform provider configuration keys for AWS credentials.
-	keyRegion                    = "region"
-	keyAccountId                 = "account_id"
-	keySessionToken              = "token"
-	keyAccessKeyID               = "access_key"
-	keySecretAccessKey           = "secret_key"
-	keyAssumeRoleWithWebIdentity = "assume_role_with_web_identity"
-	keyRoleArn                   = "role_arn"
-	keySessionName               = "session_name"
-	keyWebIdentityTokenFile      = "web_identity_token_file"
-	keyWebIdentityToken          = "web_identity_token"
-	keySkipCredsValidation       = "skip_credentials_validation"
-	keyS3UsePathStyle            = "s3_use_path_style"
-	keySkipMetadataApiCheck      = "skip_metadata_api_check"
-	keySkipRegionValidation      = "skip_region_validation"
-	keySkipReqAccountId          = "skip_requesting_account_id"
-	keyEndpoints                 = "endpoints"
+	keyAccountId = "account_id"
 )
 
 type SetupConfig struct {
 	TerraformProvider *schema.Provider
-	AWSClient         *xpprovider.AWSClient
 }
 
 func SelectTerraformSetup(config *SetupConfig) terraform.SetupFn { // nolint:gocyclo
 	return func(ctx context.Context, c client.Client, mg resource.Managed) (terraform.Setup, error) {
 		pc := &v1beta1.ProviderConfig{}
-		var err error
-		if err = c.Get(ctx, types.NamespacedName{Name: mg.GetProviderConfigReference().Name}, pc); err != nil {
+		if err := c.Get(ctx, types.NamespacedName{Name: mg.GetProviderConfigReference().Name}, pc); err != nil {
 			return terraform.Setup{}, errors.Wrapf(err, "cannot get referenced Provider: %s", mg.GetProviderConfigReference().Name)
 		}
+		t := resource.NewProviderConfigUsageTracker(c, &v1beta1.ProviderConfigUsage{})
+		if err := t.Track(ctx, mg); err != nil {
+			return terraform.Setup{}, errors.Wrap(err, "cannot track ProviderConfig usage")
+		}
+
 		ps := terraform.Setup{}
-		awsCfg, err := getAWSConfig(ctx, c, mg)
+		awsCfg, err := GetAWSConfigWithDefaultRegion(ctx, c, mg, pc)
 		if err != nil {
 			return terraform.Setup{}, errors.Wrap(err, "cannot get aws config")
```

Comment:
Agreed, but I have another suggestion.
The TF setup caller (in this case upjet connect implementation) already knows that it is calling for a terraform setup and it should be the one wrapping it with `cannot prepare the Terraform setup` (and it already does).

Here, IMHO we should give info about the "step" that TF setup failed, in order to properly distinguish and follow the errors. On top of that I suggest removing the error wrapping inside `GetAWSConfigWithDefaultRegion` implementation and let it return the downstream `GetAWSConfigWithoutTracking` error, because I realized it does not provide any extra context.